### DISCUSSION TITLE
Problem updating progress bar incrementally

### DISCUSSION THREAD

**Discussion by mlgarrett** [Textualize/textual#3135] [Q&A] [answered]
Hi, I really love Textual and I'm having a great time using it. I want to add a progress bar to display some feedback about a long-running task in my GUI. The task is a method called from an external class.

I'm working directly with the isolated progress bar example in the docs. There are two files in my example: the GUI, and the external class `object` which does an expensive task (here it just sleeps to simulate the expensive work).

I thought that a good way to do this is with a callback function that notifies the GUI to advance the progress bar each time the external class makes progress on its task. Here, I am just trying to pass the `make_progress` function in the GUI code as a callback to the external class method doing the work.

When I run this and press 'S' to start, the progress bar freezes, and then jumps to 100%. Of course, ideally, it would smoothly advance from 0 -> N and then stop. I assume this problem has something to do with asynchronicity.

I would be very appreciative if someone could point to what I need to change to achieve this correctly! I wouldn't be surprised if it's something pretty simple. Thanks in advance.

```
from textual.app import App, ComposeResult
from textual.containers import Center, Middle
from textual.timer import Timer
from textual.widgets import Footer, ProgressBar
from object import DataProcessor


class GUI(App[None]):
    BINDINGS = [("s", "start", "Start")]

    def compose(self) -> ComposeResult:
        with Center():
            with Middle():
                yield ProgressBar()
        yield Footer()

    def on_mount(self) -> None:
        self.dp = DataProcessor()

    def make_progress(self) -> None:
        self.query_one(ProgressBar).advance(1)

    def action_start(self) -> None:
        N = 10
        self.query_one(ProgressBar).update(total=N)
        self.dp.process_data(N, self.make_progress)


if __name__ == "__main__":
    GUI().run()
```

```
import time

class DataProcessor:
	def __init__(self):
		pass

	def process_data(self, N, callback):
		for i in range(N):
			# do some expensive thing
			time.sleep(0.2)
			# callback to update the progress bar
			callback()
```

**Comment by davep** [comment]
Have a look over https://textual.textualize.io/guide/workers/

**Comment by mlgarrett** [comment]
Thank you so much for the quick help!

I did look at the Workers reference, but I'm sorry to say I can't understand how to apply the Weather widget example to this. All I want is for the `process_data` function in the `object` to be able to update the progress bar in the GUI as it performs this simple iteration. As it stands now, pressing `s` in my example freezes the UI for the time it should take for the `object` to do its iteration, and then the progress bar slams to 100% in an instant.

Should the `object`'s method be `async`? Should `action_start` be `async`? What about `make_progress`? Is the callback function the right strategy here for reporting progress? Everything I have tried in the last 10 minutes hasn't worked; some have crashed, some have frozen.

**Reply by willmcgugan** [reply]
Your `process_data` method presumably takes a long time. Textual isn't able to update the UI while it is running. The solution to that is to run it concurrently (i.e. at the same time) as the UI. In Textual, Workers are the best way to manage that.

The callback in `process_data` isn't working in the way you likely intended, because the `action_start` method never exits until the entire job is done.

**Reply by willmcgugan** [reply, marked as answer]
Something like this:

```python
from textual.app import App, ComposeResult
from textual.containers import Center, Middle
from textual.timer import Timer
from textual.widgets import Footer, ProgressBar
from textual import work

import time


class GUI(App[None]):
    BINDINGS = [("s", "start", "Start")]

    def compose(self) -> ComposeResult:
        with Center():
            with Middle():
                yield ProgressBar()
        yield Footer()

    def make_progress(self) -> None:
        self.query_one(ProgressBar).advance(1)

    def action_start(self) -> None:
        N = 10
        self.query_one(ProgressBar).update(total=N)
        self.process_data(N)

    @work(thread=True)
    def process_data(self, n):
        for i in range(n):
            time.sleep(0.2)
            self.call_from_thread(self.make_progress)


if __name__ == "__main__":
    GUI().run()

```

**Reply by mlgarrett** [reply]
Thanks! This is really helpful. Forgive me if this is a stupid question, but is it possible for me to invoke the `self.call_from_thread(self.make_progress)` from the other class? In my initial example, the `process_data` function is in a separate class file, which wouldn't by default have access to the `self` of the GUI in order to update the progress bar in the GUI as it does its thing.

Again, I really appreciate the help.

**Reply by mlgarrett** [reply]
Nevermind, I got that part figured out! I can still just use a callback function. Thanks for the assistance, this is a super cool project!

**Reply by Toniolo-Marco** [reply]
How did you achieve that? I'm in the same situation, can you provide a small example?
